feat: Added support for skipping duplicates

SinTan1729 opened the pull request:
This PR adds support for skipping duplicates, closing #122 

I modified the playlist chooser dialog to add a checkbox, and then used that checkbox to decide whether duplicates should be skipped or not. It's checked by default, but it can be changed if so desired.

Before merging, the string `playlist_chooser_dialog_skip_duplicates` needs to be translated in the currently supported languages.

The toast that appears after adding songs now shows the number of songs that were actually added in the format `Added song to playlist x n` where n is the number of songs. It can be (and probably should be) changed to something more natural like `Added n song(s) to the playlist`. But that would need changes in the strings in all languages (and the position of the number might vary depending on the language, so a string replacement for a placeholder might be necessary). So I leave that decision to the maintainer.

Any kind of feedback is welcome.

eddyizm commented:
Can you add screenshots before/after and if possible an short gif of the update in action?

As far as the translations, don't worry about those, we only need the english one as they all default back to that when they don't exist. The community has been helpful with sending in those PR's otherwise we'd never make any progress. 

We may want to consider moving the checkbox to a setting depending on how busy the ui is.

SinTan1729 commented:
Thanks for the quick response. Here are the requested screenshots and recordings.

<img width="200" alt="Before" src="https://github.com/user-attachments/assets/f788623c-8902-4e2c-a933-0814b5c156c4" />


<img width="200" alt="After" src="https://github.com/user-attachments/assets/ce6821ce-a2e5-4f79-bee4-ae0c57eef54d" />

https://github.com/user-attachments/assets/48abd969-52cc-41c6-988a-dd3ed4b39d08

What's your thoughts about the toast message?

eddyizm commented:
ok, I'm taking a look now but I don't see but I'm not sure how many folks want to actually add duplicates to a playlist? 

> What are your thoughts on the toast message?  

Let's remove the count for now. I can only see the count mattering when adding an album or playlist? 
This can be tricky because of adding the count is nice but showing 0 seems like it should would be an error/warning vs reusing the success toast message. 

Would it be possible to move the checkbox to a setting, (we don't have a playlist section but i think that would be good to have now) and update the name to `allow duplicates in playlist` with the default being as you have it. 

Once we get some feedback from users on how they handle playlists, we could adjust.

eddyizm commented on app/src/main/java/com/cappielloantonio/tempo/repository/PlaylistRepository.java:
```diff
@@ -87,7 +87,7 @@ public void addSongToPlaylist(String playlistId, ArrayList<String> songsId) {
                     @Override
                     public void onResponse(@NonNull Call<ApiResponse> call, @NonNull Response<ApiResponse> response) {
-                        Toast.makeText(App.getContext(), App.getContext().getString(R.string.playlist_chooser_dialog_toast_add_success), Toast.LENGTH_SHORT).show();
+                        Toast.makeText(App.getContext(), App.getContext().getString(R.string.playlist_chooser_dialog_toast_add_success) + " x " + songsId.toArray().length, Toast.LENGTH_SHORT).show();
```
lets hold off on this, I like your other example of `Add n songs` in the near future.

SinTan1729 replied:
Once we move to prefs, I would remove this as well. Maybe something else can be done in the future.

SinTan1729 replied:
I removed this. Instead, we show a separate toast only if all songs were skipped.

eddyizm commented on app/src/main/res/values/strings.xml:
```diff
@@ -207,6 +207,7 @@
     <string name="playlist_chooser_dialog_negative_button">Cancel</string>
     <string name="playlist_chooser_dialog_neutral_button">Create</string>
     <string name="playlist_chooser_dialog_title">Add to a playlist</string>
+    <string name="playlist_chooser_dialog_skip_duplicates">Skip duplicates</string>
```
This can be used in the preferences.
```suggestion
    <string name="playlist_chooser_dialog_allow_duplicates">Allow playlist duplicates</string>
```

SinTan1729 replied:
I think saying "allow playlist duplicates" seems to suggest that Tempo will do deduplication of existing playlists. So maybe slightly changing it to "Allow adding duplicates to playlists" is better.

SinTan1729 replied:
Cool, if we move it to prefs, I'll use this wording.

eddyizm commented on app/src/main/java/com/cappielloantonio/tempo/repository/PlaylistRepository.java:
```diff
@@ -131,22 +131,22 @@ public void onFailure(@NonNull Call<ApiResponse> call, @NonNull Throwable t) {
-                    }
-                });
-    }
+//    public void updatePlaylist(String playlistId, String name, boolean isPublic, ArrayList<String> songIdToAdd, ArrayList<Integer> songIndexToRemove) {
+//        App.getSubsonicClientInstance(false)
+//                .getPlaylistClient()
+//                .updatePlaylist(playlistId, name, isPublic, songIdToAdd, songIndexToRemove)
+//                .enqueue(new Callback<ApiResponse>() {
+//                    @Override
+//                    public void onResponse(@NonNull Call<ApiResponse> call, @NonNull Response<ApiResponse> response) {
+//
+//                    }
+//
+//                    @Override
+//                    public void onFailure(@NonNull Call<ApiResponse> call, @NonNull Throwable t) {
+//
+//                    }
+//                });
+//    }
```
OK let me look for references or when it was implemented. If it is actually dead we should remove instead of commenting out.

SinTan1729 replied:
I removed it. Everything works well, and there were no references to it anywhere in the code.

SinTan1729 commented:
I can move it to playlists if you want, but all the apps I know of that have this feature (e.g. Feishin, Synfonium) do it similarly what I've done here i.e. having a checkbox/toggle in the playlist chooser screen/dialog. It doesn't mean that we can't do it differently, but we'll probably have a good overlapping of users with Feishin so it might be good to keep in mind. 

Anyway, as I said, I have no problem moving it to preferences of you so prefer.

SinTan1729 commented:
Hi, is there any updates regarding what you want me to do with this? I can either keep it as it is (which will be similar to how other apps do it), or move the config into preferences if that is preferred.

eddyizm commented:
> Hi, are there any updates regarding what you want me to do with this? I can either keep it as it is (which will be similar to how other apps do it), or move the config into preferences if that is preferred.

I haven't had time to test this out locally. There's a few other PR's that I wanted to get in and needed help. I left one comment but I think if we move this to preference I would be happier. I have used a handful of apps but not the ones you mentioned. 

I really don't see some place where a user is going to edit a playlist and on the fly change their mind to allow dupes or not allow dupes, varying by playlist to playlist? I basically would have it off and not allow it and never change it myself so I don't want it taking up precious real estate that I could use for playlist management. 

I wish I knew if there were more folks who actually use it this way, is that what you do? 

I also have 3 other requests above that need to be resolved. Happy to go over them with you but if you can do that and move it to preference, or perhaps give me a stronger case  besides other apps do it the same way, I may reconsider - otherwise, I will make sure to get this change into the next release.

SinTan1729 commented:
I have moved the config to preferences as per your suggestion. I have also removed the number from the toast. However, we now show a separate toast if all songs were skipped, so that the user is not confused. Also, the toasts were in singular, so I added `(s)` to them (i.e. `song(s)` instead of `song`).

eddyizm commented:
ok, resolved the conflicts and had a build issue, fixed the build issue but for some reason I can't push up this branch. I left my changes in a [draft PR](https://github.com/eddyizm/tempo/pull/149) if you want to take a cherry pick to merge development in with the fixes. 
Just make sure if you fix it here that it also builds successfully. There was a bad string name introduced in the polish language translation. 

Everything looks good functionally otherwise and I will get this in the next release!
